### PR TITLE
Multi page website

### LINKED ISSUE #1: Setup multi-page react website
Initial thoughts on page structure:

- / Home (Brands the website itself, variation on Render except content is flat)
- /edit Editor (form) See #9 
- /display Display (avoid 'render'). Pieces could be renamed from MainPage, after code layout is decided. To avoid nesting imports, all custom or form logic could be in one file.
- /about About - keep this minimal maybe and link to repo or docs site.
- /showcase Showcase - Collected designs.
- /terms Terms of use. See #2 

If content is mostly static, they should still be react pages for the menu.

### PR DESCRIPTION
**Pull request by akshay-nm**
Resolve #1 

### Checklist (Ticks suggest that apt components have been added)
- [x] Setup basic layout with routes for all pages and navigation
- [x] Home page 
- [x] Custom page 
- [x] Showcase page 
- [x] About page
- [x] Terms page

This does not mean there is actual content on those components. Actual content can be and probably should be added on separate PRs.

### CODE REVIEW

**Comment by akshay-nm** [comment]
Would you be comfortable if I added redux for local state management?
I also don't think we would need to use `history` module the way it's being used at the moment. I can add react hooks for the same.

File: src/App.js [hunk 1]
```diff
@@ -10,9 +18,17 @@ class App extends Component {
 
     render() {
         return (
-            <main role="main">
-                <MainPage />
-            </main>
+          <Router>
+            <Header />
+            <Switch>
+              <Route exact path='/'><Redirect to='/home' /></Route>
```
Comment (MichaelCurrin) [inline src/App.js]: Can we collapse the `/` and `/home` routes to just `/`? Thanks

**Comment by MichaelCurrin** [comment]
> Would you be comfortable if I added redux for local state management?

Sure. Just add a note in one of the docs at the bottom of the page explaining why redux is needed or how it is used.

https://github.com/MichaelCurrin/instant-website/blob/master/docs/installation.md

> I also don't think we would need to use `history` module the way it's being used at the moment. I can add react hooks for the same.

`history` came with the example I started out with and there's a comment about it here. If you can keep the existing behavior where the URL address bar updates based on values inputted in the form, then you can take `history` out.

```javascript
// Link the history used in our app to url-query so it can update the URL with it.
configureUrlQuery({ history });
```

**Comment by akshay-nm** [comment]
I have successfully replicated the current behavior. I will be adding notes regarding the same. But all in all, the custom page is complete. Requesting a review @MichaelCurrin

**Comment by akshay-nm** [comment]
I have also added components for other pages. I would need content if I am to work on them.

**Comment by MichaelCurrin** [comment]
Thanks. Its weekend now so am looking at this properly

File: src/Components/custom/Custom.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  useEffect(() => {
+    if(typeof editParam === 'number') dispatch(isDataBeingEditedUpdated(editParam == 1 ? true : false))    
+    if(titleParam) dispatch(titleUpdated(titleParam))
+    if(subtitleParam) dispatch(subtitleUpdated(subtitleParam))
+    if(descriptionParam) dispatch(descriptionUpdated(descriptionParam))
+    if(mainImageParam.length > 0) dispatch(mainImageKeywordsChanged(mainImageParam.join(',')))
+    if(bgImageParam.length > 0) dispatch(bgImageKeywordsChanged(bgImageParam.join(',')))
+
+    if(typeof editParam === 'number') console.log('redux state updated from editParam')    
+    if(titleParam) console.log('redux state updated from titleParam')
```
Comment (MichaelCurrin) [inline src/Components/custom/Custom.js]: I am happy for these to stay in permanently 
Maybe as `console.debug` ?

File: src/Components/custom/Custom.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  const [titleParam, setTitle] = useQueryParam('title', StringParam)
+  const [subtitleParam, setSubtitle] = useQueryParam('subtitle', StringParam)
+  const [descriptionParam, setDescription] = useQueryParam('description', StringParam)
+  const [mainImageParam, setMainImage] = useQueryParam('mainImage', withDefault(ArrayParam, []))
+  const [bgImageParam, setBgImage] = useQueryParam('bgImage', withDefault(ArrayParam, []))
+
+  const dispatch = useDispatch()
+
+  useEffect(() => {
+    if(typeof editParam === 'number') dispatch(isDataBeingEditedUpdated(editParam == 1 ? true : false))
```
Comment (MichaelCurrin) [inline src/Components/custom/Custom.js]: I am not a fan of one line if statements. Can these split out as two lines.
ideally managed with eslint config if that supports it

**Review by MichaelCurrin — CHANGES_REQUESTED** [review]
Thanks looks great. I made some comments and also I still need to check the frontend

File: src/Components/custom/Custom.js (hunk 2 above)
Comment (akshay-nm) [reply src/Components/custom/Custom.js]: Done

File: src/Components/custom/Custom.js (hunk 3 above)
Comment (akshay-nm) [reply src/Components/custom/Custom.js]: Done

File: src/App.js (hunk 1 above)
Comment (akshay-nm) [reply src/App.js]: Done

**Comment by akshay-nm** [comment]
@MichaelCurrin I added suggested changes. Let me know if you need anything else.

**Comment by MichaelCurrin** [comment]
Thanks @mundane140 

There can be separate PRs for content as you suggest.

I'll add this to a develop branch on my repo. Can I just request a fix to the Custom page? Previously the width was set that it the card would take up about a 1/3 of the page width on desktop (and most of the width on mobile) but now it's look around 3/4 of the width.

Can you set it to how it was before?

Previous

Screen Shot 2020-06-30 at 11 29 15 am

Current

Screen Shot 2020-06-30 at 11 29 23 am

**Comment by akshay-nm** [comment]
@MichaelCurrin I fixed the width of the card on the custom page as requested. I also fixed the warnings due to some unused imports.

**Review by MichaelCurrin — APPROVED** [review]

**Comment by MichaelCurrin** [comment]
Great, thanks for the changes. I am merging into develop.

A good follow-up PR if you feel like contributing more would be #47 which I just created.

File: src/Components/custom/View.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+          <p className="content">{description || MAIN_PAGE_DISPLAY_DEFAULTS.description}</p>
+          <br />
+
+          <footer className="card-footer">
+            <p className="card-footer-item">
+              <span>
+                Made with <a href={REPO_URL}>{PROJECT_TITLE}</a>
+                {refreshMsg}
+                <button className='button' onClick={onEditPageButtonClick}>Edit this website</button>
```
Comment (MichaelCurrin) [inline src/Components/custom/View.js]:
```suggestion
                <button className='button' onClick={onEditPageButtonClick}>Edit this page</button>
```

File: src/Components/custom/View.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          </div>
+
+          <p className="content">{description || MAIN_PAGE_DISPLAY_DEFAULTS.description}</p>
+          <br />
+
+          <footer className="card-footer">
+            <p className="card-footer-item">
+              <span>
+                Made with <a href={REPO_URL}>{PROJECT_TITLE}</a>
+                {refreshMsg}
```
Comment (MichaelCurrin) [inline src/Components/custom/View.js:47]:
```suggestion
                {refreshMsg}
                <br>
```

File: src/Components/custom/Custom.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    if(mainImageParam.length > 0) console.debug('redux state updated from mainImageParam')
+    if(bgImageParam.length > 0) console.debug('redux state updated from bgImageParam')
+  }, [])
+
+  console.debug(bgImageUrl)
+  return (
+    <div style={{ width: '100%', height: '100%', position: 'absolute', overflow: 'auto', backgroundImage: `url(${bgImageUrl})` }}>
+      <div className='columns'>
+        <div className='column'></div>
+        <div className='column'>
```
Comment (MichaelCurrin) [inline src/Components/custom/Custom.js:45]: This is okay as a 3-column approach but the modal approach of before was nice as I could set width to be more than a 3rd of the page.

```
max-width: 80%;
```

**Comment by MichaelCurrin** [comment]
Great, thanks for the changes and all the work. I'm merging to `develop`.

I have a couple of projects on the go at the same time so I can't give more attention to this one at the moment.

**Pull request merged**